code/modules/virus2/centrifuge.dm

```diff
@@ -30,114 +46,259 @@
+		return FALSE
+
 	if(.)
 		return
-	if(!is_operational())
-		return FALSE
-	if(!istype(I, /obj/item/weapon/reagent_containers/glass/beaker/vial) || !iscarbon(user))
+
+	if (istype(I,/obj/item/weapon/reagent_containers/glass/beaker/vial))
+		special = 0
+		if (on)
+			to_chat(user,"<span class='warning'>You cannot add or remove vials while the centrifuge is active. Turn it Off first.</span>")
+			return
+		var/obj/item/weapon/reagent_containers/glass/beaker/vial/vial = I
+		for (var/i = 1 to vials.len)
+			if(!vials[i])
+				vials[i] = vial
+				vial_valid[i] = vial_has_antibodies(vial)
+				visible_message("<span class='notice'>\The [user] adds \the [vial] to \the [src].</span>","<span class='notice'>You add \the [vial] to \the [src].</span>")
+				playsound(loc, 'sound/machines/click.ogg', 50, 1)
+				user.drop_item(vial, loc, 1)
+				vial.forceMove(src)
+				update_icon()
+				updateUsrDialog()
+				return TRUE
+
+		to_chat(user,"<span class='warning'>There is no room for more vials.</span>")
 		return FALSE
 
-	var/mob/living/carbon/C = user
-	if(!sample)
-		if(!C.drop_item(I, src))
-			return FALSE
-		sample = I
 
-	updateUsrDialog()
-	return TRUE
+/obj/machinery/disease2/centrifuge/proc/vial_has_antibodies(var/obj/item/weapon/reagent_containers/glass/beaker/vial/vial)
+	if (!vial)
+		return FALSE
+
+	var/datum/reagent/blood/blood = locate() in vial.reagents.reagent_list
+	if (blood && blood.data && blood.data["immunity"])
+		var/list/immune_system = blood.data["immunity"]
+		if (istype(immune_system) && immune_system.len > 0)
+			var/list/antibodies = immune_system[2]
+			for (var/antibody in antibodies)
+				if (antibodies[antibody] >= 30)
+					return TRUE
 
 //Also handles luminosity
-/obj/machinery/centrifuge/update_icon()
-	if(stat & BROKEN)
-		icon_state = "[base_state]b"
-		set_light(0)
-	else if(stat & NOPOWER)
-		icon_state = "[base_state]0"
+/obj/machinery/disease2/centrifuge/update_icon()
+	overlays.len = 0
+	icon_state = "centrifuge"
+
+	if (stat & (NOPOWER))
+		icon_state = "centrifuge0"
+
+	if (stat & (BROKEN))
+		icon_state = "centrifugeb"
+
+	if(stat & (BROKEN|NOPOWER))
 		set_light(0)
-	else if(isolating || curing)
-		set_light(l_range = 2, l_power = 2, l_color = LIGHT_COLOR_CYAN)
-		icon_state = "[base_state]_moving"
 	else
-		icon_state = "[base_state]"
-		set_light(0)
+		if (on)
+			icon_state = "centrifuge_moving"
+			set_light(2,2)
+			var/image/centrifuge_light = image(icon,"centrifuge_light")
+			centrifuge_light.plane = LIGHTING_PLANE
+			centrifuge_light.layer = ABOVE_LIGHTING_LAYER
+			overlays += centrifuge_light
+			var/image/centrifuge_glow = image(icon,"centrifuge_glow")
+			centrifuge_glow.plane = LIGHTING_PLANE
+			centrifuge_glow.layer = ABOVE_LIGHTING_LAYER
+			centrifuge_glow.blend_mode = BLEND_ADD
+			overlays += centrifuge_glow
+		else
+			set_light(2,1)
+
+		if (special == 1)
```

Comment:
Detail at this point, but this is a magic number

Reply:
added defines